File: algorithm/leetcodeCppTmp/kth-smallest-element-in-a-bst.cpp
// 230. Kth Smallest Element in a BST
// Difficulty: Medium

// Given a binary search tree, write a function kthSmallest to find the kth smallest element in it.

// Note: 
// You may assume k is always valid, 1 ≤ k ≤ BST total elements.

// Follow up:
// What if the BST is modified (insert/delete operations) often and you need to find the kth smallest frequently? 
// How would you optimize the kthSmallest routine?

// Hint:
// Try to utilize the property of a BST.
// What if you could modify the BST node structure?
// The optimal runtime complexity is O(height of BST).
 
// Time:  O(max(h, k))
// Space: O(h)
class Solution2 {
public:
    int kthSmallest(TreeNode* root, int k) {
        stack<TreeNode *> s;
        TreeNode *cur = root;
        int rank = 0;
        while (!s.empty() || cur) {
            if (cur) {
                s.emplace(cur);
                cur = cur->left;
            } else {
                cur = s.top();
                s.pop();
                if (++rank == k) {
                    return cur->val;
                }
                cur = cur->right;
            }
        }

        return INT_MIN;
    }
};

File: data-structure-and-algorithm/javascript/leetcode/kth-smallest-element-in-a-bst.js
/*
Given the root of a binary search tree, and an integer k, return the kth (1-indexed) smallest element in the tree.

 

Example 1:


Input: root = [3,1,4,null,2], k = 1
Output: 1
Example 2:


Input: root = [5,3,6,2,4,null,null,1], k = 3
Output: 3
 

Constraints:

The number of nodes in the tree is n.
1 <= k <= n <= 104
0 <= Node.val <= 104
 

Follow up: If the BST is modified often (i.e., we can do insert and delete operations) and you need to find the kth smallest frequently, how would you optimize?
*/

function kthSmallest(root, k) {
    const left = dfs(root.left);
    if (left == k - 1) {
        return root.val;
    }
    if (left >= k) {
        return kthSmallest(root.left, k);
    }
    return kthSmallest(root.right, k - left - 1);
}

function dfs(node){
    if (!node) {
        return 0;
    }
    return 1 + dfs(node.left) + dfs(node.right);
}